Expose `Extensions` during ser+de through `ron::Options`

MomoLangenstein opened the pull request:
This PR is based on #281 and #339 (with a potential application for #334 as well).

1) It adds a new `Options` struct from which all ser+de utils methods can be called.
2) The existing utils methods `XXX` now use `Options::default().XXX` as their implementation.
3) The `Options` struct can be configured with a set of default RON extensions
  a) During deserialization default extensions are automatically enabled, i.e. do not have to be included in the RON.
  b) During serialization default extensions are automatically enabled and are **not** included in the output RON. Additional extensions enabled through the `PrettyConfig` which are not in the set of default extensions are still included in the output RON. Therefore, this PR supersedes #339.
  c) This now enables programs which always use a set of extensions to enable them programmatically and exclude them from the RON config files.
4) This PR is fully backwards compatible by adding one extra method each to the `Serializer` and `Deserializer` and not changing the default behaviour (i.e. no default extensions).
5) This PR also does a tiny bit of cleanup in the docs and `Default` impl for `PrettyConfig`.

* [x] Add and update the documentation
* [x] I've included my change in `CHANGELOG.md`

MomoLangenstein commented:
@torkleyy Here's my initial attempt at the `Option` - I'm looking forward to your feedback!

torkleyy commented on src/options.rs:
```diff
@@ -0,0 +1,117 @@
+
+    #[must_use]
+    pub fn with_default_extensions(mut self, default_extensions: Extensions) -> Self {
+        self.default_extensions = default_extensions;
```
```suggestion
        self.default_extensions |= default_extensions;
```

MomoLangenstein replied:
That would definitely allow for nicer code - do you think it would then also need a separate replacement / reset method?

torkleyy replied:
I'm not sure why that would be needed. Unless you got a specific case in mind, I wouldn't add that.

MomoLangenstein replied:
Ok, thanks for your assessment. I didn't have anything in mind really, my usecase has a constant config that I only need to build once.

torkleyy replied:
If the extensions are just assigned, there's no need to make use of the builder pattern imo - I guess we could offer both, a builder and a "constructor".

MomoLangenstein replied:
I agree that right now a builder is certainly overkill. But maybe additional config options will be added in the future, when having a constructor with lots of parameters would become annoying.

torkleyy reviewed (commented):
Sorry for the late review. I like where this is going!

MomoLangenstein commented:
> Sorry for the late review. I like where this is going!

@torkleyy If you like the API as it is right now (and you give me the go-ahead), I'll start updating the documentation.

torkleyy reviewed (approved):
Looks great, go ahead :+1:

MomoLangenstein commented:
@torkleyy I've added documentation and the CHANGELOG entry. I also made the following changes to the `Options` API:
* `Options::build()` is now `Options::default()`
* both `Serializer` and `Deserializer` now take the `Options` in their new constructors instead of the `default_extensions`. This makes `Options` more obtrusive but makes it much easier to add new options to the `Options`, e.g. for #334

I've also made some minor code cleanup adjustments:
* moved the `Position` from `parse` to `errors` as it is exposed anyways (both in `de` and `error` and rustdoc was not showing them as the exact same struct)
* made `ser::State` private
* hid the doc for `ser::Compound` as it is an implementation detail but cannot be private
* simplified `PrettyConfig`'s `Default` impl so that it isn't duplicated between serde and `Default`

torkleyy commented on src/ser/mod.rs:
```diff
@@ -77,25 +70,18 @@ struct Pretty {
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct PrettyConfig {
     /// Limit the pretty-ness up to the given depth.
-    #[serde(default = "default_depth_limit")]
```
I'm not sure how this is meant to work, if you remove these serde attributes it will default to 0 or empty strings.

MomoLangenstein replied:
You are absolutely right, I forgot to add `#[serde(default)]` (https://serde.rs/container-attrs.html#default) to the struct. Should be fixed now

torkleyy replied:
Oh, interesting, did not know about the container-level default attribute.

torkleyy commented on src/options.rs:
````diff
@@ -0,0 +1,146 @@
+/// ```
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(default)]
+pub struct Options {
````
We could add `non_exhaustive` for this struct.

MomoLangenstein replied:
Thank you for that suggestion! I've used the same approach as in `PrettyConfig` to fix this
